R/methods-CollectionWithMetadata.R

```diff
@@ -0,0 +1,247 @@
+    if (!inherits(dt, 'data.table')) {
+        data.table::setDT(dt)
+    }
+
+    if (asCopy) {
+        dt <- data.table::copy(dt)
+    }
+
+    if (object@removeEmptyRecords) {
+        # not using getCollectionValues here bc i want the empty records here
+        abundances <- object@data[, -..allIdColumns]
+
+        # Remove metadata for records with NA or 0 in all columns
+        dt <- dt[rowSums(isNAorZero(abundances)) != ncol(abundances),]
+    }
+
+    if (includeIds && !is.null(metadataVariables)) {
+        dt <- dt[, c(allIdColumns, metadataVariables), with = FALSE]
+    } else if (!includeIds && !is.null(metadataVariables)) {
+        dt <- dt[, metadataVariables, with = FALSE]
+    } else if (!includeIds && is.null(metadataVariables)) {
+        dt <- dt[, -..allIdColumns]
+    }
+
+    return(dt)
+})
+
+#' Drop Records with incomplete SampleMetadata
+#'
+#' Modifies the data and sampleMetadata slots of an 
+#' CollectionWithMetadata object, to exclude records with 
+#' missing SampleMetadata for a specified column.
+#' 
+#' @param object CollectionWithMetadata
+#' @param colName String providing the column name in SampleMetadata to check for completeness
+#' @param verbose boolean indicating if timed logging is desired
+#' @return CollectionWithMetadata with modified data and sampleMetadata slots
+#' @rdname removeIncompleteRecords
+#' @export
+setGeneric("removeIncompleteRecords",
+    function(object, colName = character(), verbose = c(TRUE, FALSE)) standardGeneric("removeIncompleteRecords"),
+    signature = c("object")
+)
+
+#' @rdname removeIncompleteRecords
+#' @aliases removeIncompleteRecords,CollectionWithMetadata-method
+setMethod("removeIncompleteRecords", signature("CollectionWithMetadata"), function(object, colName = character(), verbose = c(TRUE, FALSE)) {
+    verbose <- veupathUtils::matchArg(verbose)
+
+    df <- getCollectionValues(object, verbose = verbose)
+    sampleMetadata <- getSampleMetadata(object)
+
+    # Remove Records with NA from data and metadata
+    if (any(is.na(sampleMetadata[[colName]]))) {
+        veupathUtils::logWithTime("Found NAs in specified variable. Removing these records.", verbose)
+        recordsWithData <- which(!is.na(sampleMetadata[[colName]]))
+        # Keep records with data. Recall the CollectionWithMetadata object requires records to be in the same order
+        # in both the data and metadata
+        sampleMetadata <- sampleMetadata[recordsWithData, ]
+        df <- df[recordsWithData, ]
+
+        object@data <- df
+        object@sampleMetadata <- SampleMetadata(
+            data = sampleMetadata,
+            recordIdColumn = object@sampleMetadata@recordIdColumn
+        )
+        validObject(object)
+    }
+
+    return(object)
+})
+
+#' Prune features by predicate
+#' 
+#' Modifies the data slot of an 
+#' CollectionWithMetadata object, to exclude features for which 
+#' the provided predicate function returns FALSE.
+#' 
+#' @param object CollectionWithMetadata
```

Comment:
This could run on the more general Collection, no?

Reply:
yea i suppose it could